### ISSUE TITLE
HTTP getAttachment broken in Electron ( nodeIntegration: true )

### ISSUE THREAD

**Issue by knubie** [pouchdb/pouchdb#7836]
### Issue
When calling `db.getAttachment` in the Electron renderer process, PouchDB incorrectly guesses that the function is being called in a node (or non-browser environment) and calls `response.buffer()` instead of `response.blob()`. `buffer` is undefined and raises an error. (Incidentally, this error message is not bubbled up to the promise, which made it difficult to debug).

The relevant code can be found here: https://github.com/pouchdb/pouchdb/blob/160f481e08d052919219e0115b6e65034303d7ec/packages/node_modules/pouchdb-adapter-http/src/index.js#L622-L627

Where `process` is defined, but `process.browser` is not.

**NOTE**: I believe that `nodeIntegration` must be set to true in the `BrowserWindow` options for `process` to be defined in the renderer process. In Electron <5.x.x  this value is set to `true` by default.

#### Workaround
At the beginning of your code you can add `process.browser = true` to work around the issue.

### Info
- Environment: browser
- Platform: Electron
- Adapter: IndexedDB
- Server: CouchDB

### Reproduce
- Create a new electron process.
- Create two a CouchDB database with a document with at least one attachment.
- Call `getAttachment` from the renderer process.

**Comment by MichaelJCole** [comment]
Wild-ass guess.  Try the browser version of pouchdb: `npm install pouchdb-browser` instead of `npm install pouchdb`

**Comment by DarrenCook** [comment]
Thanks so much, for that simple workaround! I've been collecting similar bug reports as I try to troubleshoot this since yesterday afternoon, and don't know how I missed your bug report until now.

BTW, I was getting this yesterday in chromium (Quasar, mode SPA) on first load, but then it always worked after a hot reload. I.e. not just electron (but with electron a reload never helped). However I haven't been able to reproduce it today.

https://stackoverflow.com/q/54595908/841830  Not Electron-specific, and his workaround was a nasty hack, and then his other workaround (https://stackoverflow.com/q/54595908/841830) looked angular specific. But I think this is the same bug, and I suspect `process.browser=true` would fix it there too.

https://stackoverflow.com/q/51672227/841830 looks to be the same (but connecting to Cloudant not CouchDB, and again using angular, not electron).

Same error message, but probably different cause: https://github.com/pouchdb-community/pouchdb-replication-stream/issues/58

Then there was this https://stackoverflow.com/q/53416865/841830, but it was fixed by unblocking a firewall port. However lots of troubleshooting about the frustratingly blank error message at https://github.com/pouchdb/pouchdb/issues/7450  (oh, the OP there *is* using Electron, so probably his was the same problem?)

**Comment by DarrenCook** [comment]
@MichaelJCole I just did `yarn remove pouchdb`, then `yarn add pouchdb-browser`, and commented out the `process.browser=true` hack. At first it seemed to be working, and I got all excited... but I must have had something cached, as when I made sure I had cleared out all saved config, and made a fresh electron build, it failed.
And then uncommenting `process.browser=true` (and again making a fresh build) fixed it.

So, sorry, I think the wild guess is wrong :-)

**Comment by DarrenCook** [comment]
A few more notes I made, that are probably not important:

1. This warning appears in the electron console log, with or without `process.browser=true`, and whether replication works or not. I.e. it is a red herring (I've still not tracked down which code is triggering it.)

>internal/process/warning.js:23 (node:25965) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.

2. If using `quasar dev -m electron` to test the electron build, if I comment out `process.browser = true` and run it again, it will still work. (My guess is the data is now in IndexedDB, so it does not need to attempt the replication?) So, to make sure testing is clean, on Linux, delete the `~/.config/Electron` directory. When testing with a production build, delete the config directory for the name of your app.

3. I'm using Electron 5.0.7. If I try switching `nodeIntegration: true` to `false`, I have other things break, so I cannot say if doing that will fix the problem or not.

**Comment by DarrenCook** [comment]
Dear StaleBot, please don't mark this as wontfix, let alone closed. As my list of links shows, this is a widespread problem, probably caused by making the false assumption that if `process` is defined that you should `response.buffer()` instead of `response.blob()`.

I.e. assuming that `process` is set **and only set** when running in Node. This is invalid logic: `process` is set in some non-Node environments, such as the renderer processes in an Electron app.

**Comment by DarrenCook** [comment]
Dear Stale Bot Owners,
Can we put the dog on a tighter leash? Bugs don't magically get fixed as they get older. And relatively few people reporting on it does not mean it is not a real bug, or does not affect people.

AFAIK, this is definitely a bug, a bad assumption, which has a serious effect on a relatively unusual configuration. (I.e. any non-Node application that defines `process`, and is using attachments).

Just because the contributors to this issue so far only know a hacky workaround, and not how to do a proper fix, does not qualify it as "won't fix". It should be kept open, for years and years if necessary, until someone comes along with the magical combination of both the need and the knowledge to fix it.

**Comment by knubie** [comment]
@DarrenCook  actually I think this was fixed with #7933.

**Comment by DarrenCook** [comment]
@knubie Thanks, I'd missed that!
I skimmed recent commits to see what else I might have missed, and noticed this one, by @garfieldnate: 
 https://github.com/pouchdb/pouchdb/commit/8351070a2f6a04c4d815f486ba460ace04f459e3

As it replaces the `if (typeof process !== 'undefined' && !process.browser)` with different code, whereas you added `&& typeof response.buffer === 'function'`, I wondered if they might be fixing the same kind of bug, but in different ways? And if so, is one better than the other, and should they be consistent?

**Comment by daleharvey** [comment]
Bugs get magically fixed magically all the time, electron could stop referencing process, or people could stop using electron, or people could fix the bug in a different issue and not update this one which seems to be exactly whats happened here, so cheers Stalebot

**Comment by garfieldnate** [comment]
If we are checking if we can call a method `buffer`, then I don't think we still need the check for `process`. In general, I think it's better to check if features we want can be used, rather than using heuristics to figure out the environment and then act based on hardcoded information about what the particular environment is capable of at the time of writing.

I received the same comment on my PR, which I then updated to follow this advice, but I can't find the PR now.

As for the two alternative methods, `response.buffer === 'function'` vs `'buffer' in response`: I think the first one is more specific and therefore better, but I don't know if it will make a big difference in practice. (Who would be annoying enough to create a new JS environment where `response` has a `buffer` field but it's not a function?) Uniting them is a great idea though. Create a method like `hasBufferFunction` and use it in the two places in the file, or throughout the library. I suspect that a library this size could use an entire utilities file for checking environment capabilities.
